plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java

```diff
@@ -552,6 +556,7 @@ private void invalidateTableCaches(SchemaTableName schemaTableName)
         // TODO https://github.com/trinodb/trino/issues/12526: invalidate tableHandlesByNameCache for handles derived from opaque queries
         invalidateColumnsCache(schemaTableName);
         invalidateCache(tableHandlesByNameCache, key -> key.tableName.equals(schemaTableName));
+        tableHandlesByQueryCache.invalidateAll();
```

Comment:
This looks too coarse. For every table invalidation (with any cause) all caches for handles by query are invalidated.

Apart from that I've noticed that the only usage of 
`BaseJdbcClient.getTableHandle(ConnectorSession, PreparedQuery)  (io.trino.plugin.jdbc)` 
is for 
`visitTableFunctionInvocation(TableFunctionInvocation, Optional<Scope>) in Visitor in StatementAnalyzer  (io.trino.sql.analyzer)`

![image](https://user-images.githubusercontent.com/3051719/197735321-946857db-8be1-4bac-8059-dcdb3f1dfd2d.png)

So this test (evolved since initial PR) is not really precise as it tests real tables (but should not), but not the virtual ones.
```
        dropTable(phantomTable); // not via CachingJdbcClient

        assertThatThrownBy(() -> jdbcClient.getTableHandle(SESSION, query))
                .hasMessageContaining("Failed to get table handle for prepared query");

        assertCacheStats(cachingJdbcClient, TABLE_HANDLES_BY_QUERY_CACHE)
                .hits(1)
                .afterRunning(() -> {
                    assertThat(cachingJdbcClient.getTableHandle(SESSION, query))
                            .isEqualTo(cachedTable);
                    assertThat(cachingJdbcClient.getColumns(SESSION, cachedTable))
                            .hasSize(0); // phantom_table has no columns
                });

// ...

        cachingJdbcClient.createTable(SESSION, new ConnectorTableMetadata(phantomTable, emptyList()));

        assertCacheStats(cachingJdbcClient, TABLE_HANDLES_BY_QUERY_CACHE)
                .misses(1)
                .loads(1)  //                          <<<------------------ this has to be 0, and not tested on real table
                .afterRunning(() -> {
                    assertThat(cachingJdbcClient.getTableHandle(SESSION, query))
                            .isEqualTo(cachedTable);
                    assertThat(cachingJdbcClient.getColumns(SESSION, cachedTable))
                            .hasSize(0); // phantom_table has no columns
                });
```

WDYT?

Reply:
> This looks too coarse. For every table invalidation (with any cause) all caches for handles by query are invalidated.

Can we do better?

> So this test (evolved since initial PR) is not really precise as it tests real tables (but should not), but not the virtual ones.

i am not sure i follow. Do you want to make a PR with a fix?

Reply:
I'm trying figuring out what's going on and this is my findings. Actually my thoughts. 
Looking for any comments which would allow me to shape that knowledge.